Sources/Testing/Parameterization/CustomTestArgument.swift

```diff
@@ -0,0 +1,154 @@
+  func encodeTestArgument(to encoder: some Encoder) throws
+}
+
+extension Test.Case.Argument.ID {
+  /// Initialize this instance with an ID for the specified test argument.
+  ///
+  /// - Parameters:
+  ///   - value: The value of a test argument for which to get an ID.
+  ///   - parameter: The parameter of the test function to which this argument
+  ///     value was passed.
+  ///
+  /// - Returns: `nil` if an ID cannot be formed from the specified test
+  ///   argument value.
+  ///
+  /// - Throws: Any error encountered while attempting to encode `value`.
+  ///
+  /// This function is not part of the public interface of the testing library.
+  ///
+  /// ## See Also
+  ///
+  /// - ``CustomTestArgumentEncodable``
+  init?(identifying value: some Sendable, parameter: Test.ParameterInfo) throws {
+#if canImport(Foundation)
+    guard Configuration.current?.isTestArgumentEncodingEnabled ?? false else {
+      return nil
+    }
+
+    func _customArgumentWrapper(for value: some CustomTestArgumentEncodable) -> some Encodable {
+      _CustomArgumentWrapper(value: value)
+    }
+
+    let encodableValue: (any Encodable)? = if let customEncodable = value as? any CustomTestArgumentEncodable {
+      _customArgumentWrapper(for: customEncodable)
+    } else if let encodable = value as? any Encodable {
+      encodable
+    } else if let identifiable = value as? any Identifiable, let encodableID = identifiable.id as? any Encodable {
+      encodableID
+    } else {
+      nil
+    }
+
+    guard let encodableValue else {
+      return nil
+    }
+
+    self = .init(bytes: try Self._encode(encodableValue, parameter: parameter))
+#else
+    nil
+#endif
+  }
+
+#if canImport(Foundation)
+  /// Encode the specified test argument value and store its encoded
+  /// representation as an array of bytes suitable for storing in an instance of
+  /// ``Test/Case/Argument/ID-swift.struct``.
+  ///
+  /// - Parameters:
+  ///   - value: The value to encode.
+  ///   - parameter: The parameter of the test function to which this argument
+  ///     value was passed.
+  ///
+  /// - Returns: An array of bytes containing the encoded representation.
+  ///
+  /// - Throws: Any error encountered during encoding.
+  private static func _encode(_ value: some Encodable, parameter: Test.ParameterInfo) throws -> [UInt8] {
+    let encoder = JSONEncoder()
+
+    // Keys must be sorted to ensure deterministic matching of encoded data.
+    encoder.outputFormatting.insert(.sortedKeys)
+
+    // Set user info keys which clients may wish to use during encoding.
+    encoder.userInfo[._testParameterUserInfoKey] = parameter
+
+    return .init(try encoder.encode(value))
+  }
+#endif
+}
+
+/// A encodable type which wraps a ``CustomTestArgumentEncodable`` value.
+private struct _CustomArgumentWrapper<T>: Encodable where T: CustomTestArgumentEncodable {
+  /// The value this instance wraps, which implements custom test argument
+  /// encoding logic.
+  var value: T
```

Comment:
`RawRepresentable` might be appropriate here, but it's not critical.

Reply:
Done!